File: archive/legacy-node/routes/conversations.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');
const { db } = require('../db');
const authMiddleware = require('../middleware/auth');
const { generateBotReply } = require('../services/openai');

router.use(authMiddleware);

// GET /api/conversations
router.get('/', (req, res) => {
  const { mode, status } = req.query;
  const restaurant_id = req.user.restaurant_id;

  let query = `
    SELECT conv.*, c.name as customer_name, c.platform as customer_platform,
      (SELECT content FROM messages WHERE conversation_id = conv.id ORDER BY created_at DESC LIMIT 1) as last_message,
      (SELECT created_at FROM messages WHERE conversation_id = conv.id ORDER BY created_at DESC LIMIT 1) as last_message_at
    FROM conversations conv
    JOIN customers c ON conv.customer_id = c.id
    WHERE conv.restaurant_id = ?
  `;
  const params = [restaurant_id];

  if (mode && mode !== 'all') {
    query += ' AND conv.mode = ?';
    params.push(mode);
  }

  if (status && status !== 'all') {
    query += ' AND conv.status = ?';
    params.push(status);
  }

  query += ' ORDER BY conv.updated_at DESC';

  const conversations = db.prepare(query).all(...params);
  res.json(conversations);
});

// GET /api/conversations/:id/messages
router.get('/:id/messages', (req, res) => {
  const { id } = req.params;
  const restaurant_id = req.user.restaurant_id;

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  const messages = db.prepare('SELECT * FROM messages WHERE conversation_id = ? ORDER BY created_at ASC').all(id);
  res.json(messages);
});

// POST /api/conversations/:id/messages
router.post('/:id/messages', (req, res) => {
  const { id } = req.params;
  const { content } = req.body;
  const restaurant_id = req.user.restaurant_id;

  if (!content) {
    return res.status(400).json({ error: 'Content is required' });
  }

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  const msgId = uuidv4();
  db.prepare('INSERT INTO messages (id, conversation_id, role, content) VALUES (?, ?, ?, ?)').run(
    msgId, id, 'staff', content
  );

  db.prepare('UPDATE conversations SET updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(id);

  const message = db.prepare('SELECT * FROM messages WHERE id = ?').get(msgId);
  res.status(201).json(message);
});

// PATCH /api/conversations/:id/mode
router.patch('/:id/mode', (req, res) => {
  const { id } = req.params;
  const { mode } = req.body;
  const restaurant_id = req.user.restaurant_id;

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  const newMode = mode || (conv.mode === 'bot' ? 'human' : 'bot');
  db.prepare('UPDATE conversations SET mode = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(newMode, id);

  const updated = db.prepare('SELECT * FROM conversations WHERE id = ?').get(id);
  res.json(updated);
});

// PATCH /api/conversations/:id/urgent
router.patch('/:id/urgent', (req, res) => {
  const { id } = req.params;
  const restaurant_id = req.user.restaurant_id;

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  const newUrgent = conv.urgent ? 0 : 1;
  db.prepare('UPDATE conversations SET urgent = ? WHERE id = ?').run(newUrgent, id);

  const updated = db.prepare('SELECT * FROM conversations WHERE id = ?').get(id);
  res.json(updated);
});

// PATCH /api/conversations/:id/read
router.patch('/:id/read', (req, res) => {
  const { id } = req.params;
  const restaurant_id = req.user.restaurant_id;

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  db.prepare('UPDATE conversations SET unread_count = 0 WHERE id = ?').run(id);

  const updated = db.prepare('SELECT * FROM conversations WHERE id = ?').get(id);
  res.json(updated);
});

// POST /api/conversations/:id/ai-reply
router.post('/:id/ai-reply', async (req, res) => {
  const { id } = req.params;
  const restaurant_id = req.user.restaurant_id;

  const conv = db.prepare('SELECT * FROM conversations WHERE id = ? AND restaurant_id = ?').get(id, restaurant_id);
  if (!conv) {
    return res.status(404).json({ error: 'Conversation not found' });
  }

  const restaurant = db.prepare('SELECT * FROM restaurants WHERE id = ?').get(restaurant_id);
  const messages = db.prepare('SELECT * FROM messages WHERE conversation_id = ? ORDER BY created_at ASC').all(id);

  const reply = await generateBotReply(messages, restaurant ? restaurant.name : 'Al-Rashid Kitchen');

  if (!reply) {
    return res.status(500).json({ error: 'Failed to generate AI reply' });
  }

  const msgId = uuidv4();
  db.prepare('INSERT INTO messages (id, conversation_id, role, content) VALUES (?, ?, ?, ?)').run(
    msgId, id, 'bot', reply
  );

  db.prepare('UPDATE conversations SET updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(id);

  const message = db.prepare('SELECT * FROM messages WHERE id = ?').get(msgId);
  res.status(201).json(message);
});

module.exports = router;

File: routers/conversations.py
"""
routers/conversations.py — NUMBER 43: Conversations extracted from main.py.

Routes: GET /api/conversations, GET/POST /api/conversations/{cid}/messages,
        GET /api/outbound/failed,
        PATCH /api/conversations/{cid}/messages/{mid},
        PATCH /api/conversations/{cid}/mode,
        PATCH /api/conversations/{cid}/urgent,
        PATCH /api/conversations/{cid}/read,
        POST /api/conversations/{cid}/ai-reply

Unchanged behavior — same URLs, same response shapes, same auth guards.
"""
from __future__ import annotations
import logging
import os
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel

import database
from dependencies import current_user
from helpers import log_activity

logger = logging.getLogger("restaurant-saas")

router = APIRouter()

_openai_client = None


def _get_openai_client():
    global _openai_client
    if _openai_client is None:
        api_key = os.getenv("OPENAI_API_KEY", "")
        if api_key:
            try:
                import openai as _openai
                _openai_client = _openai.OpenAI(api_key=api_key)
            except ImportError:
                pass
    return _openai_client


class MsgCreate(BaseModel):
    content: str


@router.get("/api/conversations")
async def list_conversations(
    mode: Optional[str] = None,
    status: Optional[str] = None,
    search: Optional[str] = None,
    customer_id: Optional[str] = None,
    user=Depends(current_user),
):
    conn = database.get_db()
    rid = user["restaurant_id"]
    q = """
        SELECT cv.*,
               c.name AS customer_name,
               c.platform,
               c.phone,
               COALESCE(
                 (SELECT memory_value FROM conversation_memory
                  WHERE customer_id=c.id AND restaurant_id=? AND memory_key='name'
                  ORDER BY updated_at DESC LIMIT 1),
                 c.name
               ) AS display_name,
               (SELECT content FROM messages
                WHERE conversation_id=cv.id ORDER BY created_at DESC LIMIT 1) AS last_message,
               (SELECT role FROM messages
                WHERE conversation_id=cv.id ORDER BY created_at DESC LIMIT 1) AS last_message_role,
               (SELECT created_at FROM messages
                WHERE conversation_id=cv.id ORDER BY created_at DESC LIMIT 1) AS last_message_at
        FROM conversations cv JOIN customers c ON cv.customer_id = c.id
        WHERE cv.restaurant_id=?
    """
    params = [rid, rid]
    if mode:
        q += " AND cv.mode=?"; params.append(mode)
    if status:
        q += " AND cv.status=?"; params.append(status)
    if search:
        q += """ AND (c.name LIKE ? OR c.phone LIKE ?
                   OR EXISTS(SELECT 1 FROM conversation_memory cm2
                              WHERE cm2.customer_id=c.id AND cm2.memory_key='name'
                                AND cm2.memory_value LIKE ?))"""
        params += [f"%{search}%", f"%{search}%", f"%{search}%"]
    if customer_id:
        q += " AND cv.customer_id=?"; params.append(customer_id)
    q += " ORDER BY COALESCE((SELECT created_at FROM messages WHERE conversation_id=cv.id ORDER BY created_at DESC LIMIT 1), cv.updated_at) DESC"
    rows = conn.execute(q, params).fetchall()
    conn.close()
    return [dict(r) for r in rows]


@router.get("/api/conversations/{cid}/messages")
async def get_messages(cid: str, user=Depends(current_user)):
    conn = database.get_db()
    if not conn.execute("SELECT id FROM conversations WHERE id=? AND restaurant_id=?",
                        (cid, user["restaurant_id"])).fetchone():
        conn.close()
        raise HTTPException(404, "Conversation not found")
    msgs = conn.execute(
        "SELECT * FROM messages WHERE conversation_id=? ORDER BY created_at", (cid,)).fetchall()
    conn.close()
    return [dict(m) for m in msgs]


@router.post("/api/conversations/{cid}/messages")
async def send_message(cid: str, data: MsgCreate, user=Depends(current_user)):
    conn = database.get_db()
    if not conn.execute("SELECT id FROM conversations WHERE id=? AND restaurant_id=?",
                        (cid, user["restaurant_id"])).fetchone():
        conn.close()
        raise HTTPException(404, "Conversation not found")
    mid = str(uuid.uuid4())
    conn.execute("INSERT INTO messages (id, conversation_id, role, content) VALUES (?, ?, 'staff', ?)",
                 (mid, cid, data.content))
    conn.execute("UPDATE conversations SET updated_at=CURRENT_TIMESTAMP, unread_count=0, mode='human' WHERE id=?", (cid,))
    conn.commit()
    msg = conn.execute("SELECT * FROM messages WHERE id=?", (mid,)).fetchone()
    conn.close()
    return dict(msg)


@router.get("/api/outbound/failed")
async def list_failed_messages(user=Depends(current_user)):
    conn = database.get_db()
    rows = conn.execute(
        """SELECT id, conversation_id, platform, recipient_id, content, error, created_at
           FROM outbound_messages
           WHERE restaurant_id=? AND status='failed'
           ORDER BY created_at DESC LIMIT 50""",
        (user["restaurant_id"],)
    ).fetchall()
    conn.close()
    return [dict(r) for r in rows]


@router.patch("/api/conversations/{cid}/messages/{mid}")
async def edit_message(cid: str, mid: str, req: Request, user=Depends(current_user)):
    body = await req.json()
    content = body.get("content", "").strip()
    if not content:
        raise HTTPException(400, "المحتوى مطلوب")
    conn = database.get_db()
    msg = conn.execute(
        "SELECT m.* FROM messages m JOIN conversations c ON m.conversation_id=c.id "
        "WHERE m.id=? AND m.conversation_id=? AND c.restaurant_id=?",
        (mid, cid, user["restaurant_id"])
    ).fetchone()
    if not msg:
        conn.close()
        raise HTTPException(404, "الرسالة غير موجودة")
    conn.execute("UPDATE messages SET content=? WHERE id=?", (content, mid))
    conn.commit()
    conn.close()
    return {"ok": True}


@router.patch("/api/conversations/{cid}/mode")
async def toggle_mode(cid: str, req: Request, user=Depends(current_user)):
    body = await req.json()
    mode = body.get("mode", "bot")
    conn = database.get_db()
    if not conn.execute("SELECT id FROM conversations WHERE id=? AND restaurant_id=?",
                        (cid, user["restaurant_id"])).fetchone():
        conn.close()
        raise HTTPException(404, "Conversation not found")
    conn.execute("UPDATE conversations SET mode=? WHERE id=?", (mode, cid))
    log_activity(conn, user["restaurant_id"], "conversation_mode_changed", "conversation", cid,
                 f"تغيير وضع المحادثة إلى {mode}", user["id"], user["name"])
    conn.commit()
    conn.close()
    return {"mode": mode}


@router.patch("/api/conversations/{cid}/urgent")
async def set_urgent(cid: str, req: Request, user=Depends(current_user)):
    body = await req.json()
    urgent = bool(body.get("urgent", True))
    conn = database.get_db()
    if not conn.execute("SELECT id FROM conversations WHERE id=? AND restaurant_id=?",
                        (cid, user["restaurant_id"])).fetchone():
        conn.close()
        raise HTTPException(404, "Conversation not found")
    conn.execute("UPDATE conversations SET urgent=? WHERE id=?", (int(urgent), cid))
    conn.commit()
    conn.close()
    return {"urgent": urgent}


@router.patch("/api/conversations/{cid}/read")
async def mark_read(cid: str, user=Depends(current_user)):
    conn = database.get_db()
    conn.execute("UPDATE conversations SET unread_count=0 WHERE id=? AND restaurant_id=?",
                 (cid, user["restaurant_id"]))
    conn.commit()
    conn.close()
    return {"unread_count": 0}


@router.post("/api/conversations/{cid}/ai-reply")
async def ai_reply(cid: str, user=Depends(current_user)):
    client = _get_openai_client()
    if not client:
        raise HTTPException(503, "OpenAI غير مهيأ — أضف OPENAI_API_KEY في ملف .env")
    conn = database.get_db()
    if not conn.execute("SELECT id FROM conversations WHERE id=? AND restaurant_id=?",
                        (cid, user["restaurant_id"])).fetchone():
        conn.close()
        raise HTTPException(404, "Conversation not found")
    s = conn.execute("SELECT * FROM settings WHERE restaurant_id=?", (user["restaurant_id"],)).fetchone()
    r = conn.execute("SELECT * FROM restaurants WHERE id=?", (user["restaurant_id"],)).fetchone()
    msgs = list(reversed(conn.execute(
        "SELECT * FROM messages WHERE conversation_id=? ORDER BY created_at DESC LIMIT 10", (cid,)).fetchall()))
    conn.close()

    bot_name = (s["bot_name"] if s else None) or "AI Assistant"
    rest_name = (r["name"] if r else None) or "المطعم"
    system_prompt = (
        f"أنت {bot_name}، مساعد ذكاء اصطناعي لمطعم {rest_name}.\n"
        "مهمتك مساعدة العملاء بشكل ودي ومحترف.\nأجب باختصار وبفائدة."
    )
    chat_msgs = [{"role": "system", "content": system_prompt}]
    for m in msgs:
        role = "user" if m["role"] == "customer" else "assistant"
        chat_msgs.append({"role": role, "content": m["content"]})

    resp = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=chat_msgs,
        max_tokens=300,
    )
    ai_content = resp.choices[0].message.content

    conn = database.get_db()
    mid = str(uuid.uuid4())
    conn.execute("INSERT INTO messages (id, conversation_id, role, content) VALUES (?, ?, 'bot', ?)",
                 (mid, cid, ai_content))
    conn.execute("UPDATE conversations SET updated_at=CURRENT_TIMESTAMP WHERE id=?", (cid,))
    conn.commit()
    msg = conn.execute("SELECT * FROM messages WHERE id=?", (mid,)).fetchone()
    conn.close()
    return dict(msg)
